### PR TITLE
robocup 2019 possession fixes

### LINKED ISSUE #756: Robocup 2019 possession fixes
### Description of the task

cherrypick possession fixes to master

### Outline the task in more detail

cherrypick possession fixes to master, fix existing unit tests, add additional unit tests as necessary

### Acceptance criteria

- [ ] cherrypick possession fixes
- [ ] fix existing unit tests
- [ ] add new unit tests as necessary

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

Cherrypicked the possession fixes from Robocup 2019 and fixed existing unit tests

### Testing Done

existing unit tests for changed evaluation functions

### Resolved Issues

Resolves #756

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — DISMISSED** [review]

**Review by matthewberends — DISMISSED** [review]

File: src/thunderbots/software/ai/hl/stp/evaluation/robot.h
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/robot.h]: Why are we giving this a default `timestamp`? Seems like this is something we would like to force the user to specify? Alternatively, we could make it a `std::optional<Timestamp>` and use the most recent one if the user gives `std::nullopt`? 

Similarly for `robotBeingPassedTo` below.

File: src/thunderbots/software/test/ai/hl/stp/evaluation/possession.cpp
Comment (garethellis0) [inline src/thunderbots/software/test/ai/hl/stp/evaluation/possession.cpp]: nit; no newline at EOF

**Review by garethellis0 — CHANGES_REQUESTED** [review]
I wrote 99% of this review almost week back and just failed to submit. Sorry about that, will try to be more careful about it in the future.

File: src/thunderbots/shared/constants.h
Comment (jordanzliu) [reply src/thunderbots/shared/constants.h]: This is shared for some reason. I'll move it into a DynamicParameter

File: src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp]: it was originally copypasted at some point, forgot to change the comment

File: src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp]: yeah will make the change

File: src/thunderbots/software/ai/hl/stp/evaluation/robot.h
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/evaluation/robot.h]: This was to make refactoring less of a PITA during competition. will make the change

**Comment by MathewMacDougall** [comment]
Just realized this is blocking some RoboCup changse like #888 which are pretty high-priority to unblock @jonathanlew. @jordanzliu can you have this done by the end of the weekend? Seems like it mostly needs some comments and resolved merge conflicts (which should mostly be moving files).

**Comment by jordanzliu** [comment]
Yeah I can prob get it done tomorrow. been dead with recruiting season lately

**Comment by MathewMacDougall** [comment]
that's be awesome, thanks!

**Comment by MathewMacDougall** [comment]
There's some interesting build failures in `stp/evaluation/robot.cpp` @jordanzliu

**Review by MathewMacDougall — DISMISSED** [review]
LGTM pending @garethellis0 's approval for his comments

**Comment by jordanzliu** [comment]
Still working on resolving his comments

File: src/thunderbots/software/ai/hl/stp/evaluation/possession.h
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/evaluation/possession.h]: done

**Comment by MathewMacDougall** [comment]
Yikes more conflicts (I think from #867 ) but they look easy to resolve.

**Comment by MathewMacDougall** [comment]
@jordanzliu 1 possession test failure otherwise LGTM

**Comment by jordanzliu** [comment]
Can't reproduce failure locally, will investigate more tmr night

**Comment by MathewMacDougall** [comment]
@jordanzliu any updates? I'm going to be blocked by this really soon since I need it to finish porting the plays from robocup.

**Comment by jordanzliu** [comment]
Will look at this again tonight, tried cleaning and building again to
reproduce

**Comment by MathewMacDougall** [comment]
Looks like the same failure :cry:

**Comment by jordanzliu** [comment]
Running the failed test on my machine still passes???????? This is baffling
Edit: probably undefined behaviour somewhere, will look into it more at lunch
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from RobotEvaluationTest
[ RUN      ] RobotEvaluationTest.has_possession_robot_moving_ball_in_dribbler
[       OK ] RobotEvaluationTest.has_possession_robot_moving_ball_in_dribbler (0 ms)
[----------] 1 test from RobotEvaluationTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (0 ms total)
[  PASSED  ] 1 test.
```

**Comment by garethellis0** [comment]
Just took a quick look. Issue here looks like we're not checking that the optional actually has a value. Change the test to:
```
    Point ball_position  = Point(0.07, 0);
    Vector ball_velocity = Vector(0, 0);
    Timestamp timestamp  = Timestamp::fromSeconds(0);
    Ball ball            = Ball(ball_position, ball_velocity, timestamp);

    Robot robot = Robot(0, Point(0, 0), Vector(1, 1), Angle::zero(),
                        AngularVelocity::zero(), timestamp);

    std::optional<bool> robot_has_possession = Evaluation::robotHasPossession(ball, robot);
    ASSERT_TRUE(robot_has_possession.has_value());

    EXPECT_TRUE(*robot_has_possession);
```
and it will fail, otherwise it will always pass locally. Frankly, we're lucky that this failed, this could've turned into a pretty nasty little bug.

*Note the use of `ASSERT_TRUE` when checking if the optional has a value. This is because we do not want to try to take the value of an optional if we're not sure if it has one, pretty sure that's undefined behavior*

**Comment by jordanzliu** [comment]
passed tests, fixing code coverage tonight or tmr morning

**Review by MathewMacDougall — DISMISSED** [review]
LGTM. If this could be resolved by mid-day Sunday that would be awesome so I can port the RoboCup plays. Glad that test is fixed :relieved:

**Comment by MathewMacDougall** [comment]
Looks like `DefenseShadowEnemyTacticTest.test_shadower_blocks_net_when_enemy_cannot_pass` failed. My best guess is that test behavior changed because whatever assumption was made about the enemy having possession was broken.

**Comment by jordanzliu** [comment]
fixed tests and coverage is a bit better

File: src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp
Comment (jordanzliu) [reply src/thunderbots/software/ai/hl/stp/evaluation/robot.cpp]: fixed

File: src/software/ai/hl/stp/evaluation/robot.cpp
Comment (garethellis0) [inline src/software/ai/hl/stp/evaluation/robot.cpp]: nit; missing newline at EOF

**Review by garethellis0 — COMMENTED** [review]
Just one comment, LGTM otherwise

File: src/software/ai/hl/stp/evaluation/robot.cpp
Comment (jordanzliu) [reply src/software/ai/hl/stp/evaluation/robot.cpp]: fixed

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :tada: :pray: :100:

**Comment by MathewMacDougall** [comment]
:shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit:

**Pull request merged**
